web/api/v1/api.go

```diff
@@ -1548,7 +1554,71 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
 	return apiFuncResult{nil, nil, nil, nil}
 }
 
-func (api *API) respond(w http.ResponseWriter, data interface{}, warnings storage.Warnings) {
+func (api *API) tryArrowResponse(w http.ResponseWriter, r *http.Request, data interface{}, warnings storage.Warnings) error {
+	result, ok := data.(*queryData)
+	if !ok {
+		return fmt.Errorf("invalid result data type: %T", data)
+	}
+
+	if result.ResultType != parser.ValueTypeMatrix {
+		return errors.Errorf("arrow not implemented for %s", result.ResultType)
+	}
+	matrix := result.Result.(promql.Matrix)
+
+	// Currently histograms are not handled with Arrow, just adding a binary
+	// type here halves the performance of Arrow. In the future we could
+	// consider building a custom type in Arrow for histograms.
+	for _, series := range matrix {
+		if len(series.Points) > 0 && series.Points[0].H != nil {
+			return fmt.Errorf("arrow not implemented for native histograms")
+		}
+	}
+
+	w.Header().Set("Content-Type", mimeTypeArrowStream)
+	w.WriteHeader(http.StatusOK)
+
+	pool := memory.NewGoAllocator()
+	fields := []arrow.Field{
+		{Name: "t", Type: &arrow.TimestampType{Unit: arrow.Millisecond}},
+		{Name: "v", Type: arrow.PrimitiveTypes.Float64},
+	}
+
+	for _, series := range matrix {
+		metadata := arrow.MetadataFrom(series.Metric.Map())
```

Comment:
Having worked with arrow quite a bit now, I think this is rather odd. I would have expected labels to be a column of type map, and then each row has a list of timestamps and values. That way we can also pack multiple series into a single record.